### DISCUSSION TITLE
Does the library have built-in in caching?

### DISCUSSION THREAD

**Discussion by ArystanK** [peterLaurence/MapCompose#114] [Q&A]
I'm curious if the library has built-in caching mechanics. I wrote my custom tile provider with caching (file + LRU caching), and I'm curious if I'm doing too much work and the library already has built-in caching.

**Comment by Nohus** [comment]
IIRC the tiles are cached in-memory but are not persisted to disk, so a file cache for your tile provider is useful. I myself use very long cache times, resulting in a map that works offline once you used it once.

**Comment by p-lr** [comment]
Indeed, the library isn't opinionated about how tiles are fetched. Using a particular caching policy is an implementation detail of the `TileStreamProvider`.
However, the library does use a "bitmap pooling" technique to reuse allocated memory. For example, when scrolling, some tiles become no longer visible while new ones are fetched. The library reuses the memory of no-longer visible tiles.
The benefit is less allocations, less pressure on the garbage collector, and a reduced risk of out-of-memory.
